packages/mui-base/src/ListboxUnstyled/useListbox.ts

```diff
@@ -166,6 +179,50 @@ export default function useListbox<TOption>(props: UseListboxParameters<TOption>
         event,
         props: propsWithDefaults,
       });
+
+      // Handle text navigation
+      if (event.key.length === 1) {
+        const textCriteria = textCriteriaRef.current;
+        const lowerKey = event.key.toLowerCase();
+        const currentTime = performance.now();
+        if (textCriteria.keys.length > 0) {
+          // Reset text criteria ref
+          if (
+            textCriteria.lastTime &&
+            currentTime - textCriteria.lastTime > TEXT_NAVIGATION_RESET_TIMEOUT
+          ) {
+            textCriteria.keys = [];
+            textCriteria.repeating = true;
+            textCriteria.previousKeyMatched = true;
+          } else if (textCriteria.repeating && lowerKey !== textCriteria.keys[0]) {
+            textCriteria.repeating = false;
+          }
+        }
+
+        textCriteria.lastTime = currentTime;
+        textCriteria.keys.push(lowerKey);
+
+        const textCriteriaStr = textCriteria.repeating
+          ? textCriteria.keys[0]
+          : textCriteria.keys.join('');
+        const keepFocusOnCurrent =
+          highlightedValue &&
+          !textCriteria.repeating &&
+          textCriteriaMatches(highlightedValue, textCriteriaStr, optionStringifier);
+
+        if (textCriteria.previousKeyMatched) {
+          if (!keepFocusOnCurrent) {
+            dispatch({
+              textCriteriaStr,
+              optionStringifier,
+              type: ActionTypes.textNavigation,
+              props: propsWithDefaults,
+            });
+          }
+        } else {
+          textCriteria.previousKeyMatched = false;
+        }
+      }
```

Comment:
I think we could simplify this function, now that the reducer takes care of most of the logic:

```tsx
// Handle text navigation
if (event.key.length === 1) {
  const textCriteria = textCriteriaRef.current;
  const lowerKey = event.key.toLowerCase();
  const currentTime = performance.now();
  if (
    textCriteria.searchString.length > 0 &&
    textCriteria.lastTime &&
    currentTime - textCriteria.lastTime > TEXT_NAVIGATION_RESET_TIMEOUT
  ) {
    textCriteria.searchString = lowerKey;
  } else if (
    textCriteria.searchString.length !== 1 ||
    lowerKey !== textCriteria.searchString
  ) {
    // If there is just one character in the buffer and the key is the same, do not append
    textCriteria.searchString += lowerKey;
  }

  textCriteria.lastTime = currentTime;

  dispatch({
    type: ActionTypes.textNavigation,
    textCriteriaStr: textCriteria.searchString,
    props: propsWithDefaults,
  });
}
```

Then, `textCriteria` would have to defined as such:
```tsx
const textCriteriaRef = React.useRef<TextCriteria>({
  searchString: '',
  lastTime: null,
});
```

and `textCriteriaMatches` won't need to be exported from `defaultReducer`.